src/pycram/datastructures/property.py

```diff
@@ -0,0 +1,361 @@
+        self.result = reasoning_result
+        return reasoning_result.success
+
+    @property
+    def successful(self) -> bool:
+        """
+        Returns the result of the complete tree structure. This is done by iterating over the tree and checking if all
+        nodes return True.
+
+        :return: True if all nodes in the tree are True, False otherwise
+        """
+        success = self.result
+        for node in PreOrderIter(self.root):
+            success &= node.result.success
+        return success
+
+    @property
+    def resolved(self) -> bool:
+        """
+        Returns True if all nodes in the tree are resolved. This is done by iterating over the tree and checking if all
+        nodes are either a ResolvedProperty or a PropertyOperator.
+
+        :return: True if all nodes in the tree are resolved, False otherwise
+        """
+        res = True
+        for node in PreOrderIter(self.root):
+            res &= isinstance(node, (ResolvedProperty, PropertyOperator))
+        return res
+
+
+class PropertyOperator(Property):
+    """
+    Parent class for logical operators to combine multiple properties in a tree structure. This class adds methods to
+    use Properties as part of a tree structure. Furthermore, there is a method to simplify the tree structure by merging
+    Nodes of the same type.
+    """
+
+    def __init__(self, properties: List[Property]):
+        """
+        Initialize the PropertyOperator with a list of Properties. The parent of this node is None, therefore the node is
+        always the root of the tree.
+
+        :param properties: A list of properties to which are the children of this node
+        """
+        super().__init__(parent=None, children=properties)
+        self.result = True
+
+    def simplify(self):
+        """
+        Simplifies the tree structure by merging nodes of the same type. This is done by iterating over the tree and
+        merging nodes of the same type.
+
+        :return: Returns the root node of the tree
+        """
+        for node in PreOrderIter(self.root):
+            for child in node.children:
+                if type(node) is type(child):
+                    self.merge_nodes(node, child)
+        return self.root
+
+    @staticmethod
+    def merge_nodes(node1: Node, node2: Node) -> None:
+        """
+        Merges node1 with node2 in a tree. The children of node2 will be re-parented to node1 and node2 will be deleted
+        from the tree.
+
+        :param node1: Node that should be left in the tree
+        :param node2: Node which children should be appended to node1 and then deleted
+        """
+        node2.parent = None
+        node1.children = node2.children + node1.children
+
+
+class And(PropertyOperator):
+    """
+    Class to represent a logical and operator in a tree structure. This class inherits from PropertyOperator and adds a
+    method to evaluate the children as an and operator.
```

Comment:
I am very unsure about the information thats given here. Like Inherits from is written in the class definition and the adds a method stuff down there too. Also it is required by the super class.